Задание 1
Вернемся к набору данных о видеоиграх.

Ответьте на следующие вопросы:

1) Как критики относятся к спортивным играм?
2) Критикам нравятся больше игры на PC или на PS4?
3) Критикам больше нравятся стрелялки или стратегии?

Для каждого вопроса:

сформулируйте нулевую и альтернативную гипотезы;
выберите пороговый уровень статистической значимости;
опишите полученные результаты статистического теста.

Как критики относятся к спортивным играм? 

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('video_games_sales.csv')
df

Rank                           Name                      basename  \
0          1                     Wii Sports                    wii-sports   
1          2              Super Mario Bros.              super-mario-bros   
2          3                 Mario Kart Wii                mario-kart-wii   
3          4  PlayerUnknown's Battlegrounds  playerunknowns-battlegrounds   
4          5              Wii Sports Resort             wii-sports-resort   
...      ...                            ...                           ...   
55787  55788                    Indivisible                   indivisible   
55788  55789                     Lost Ember                    lost-amber   
55789  55790                     Lost Ember                    lost-amber   
55790  55791                     Lost Ember                    lost-amber   
55791  55792                     Falcon Age                    falcon-age   

                  Genre ESRB_Rating Platform         Publisher  \
0                Sports           E      Wii          Nintendo   
1              Platform         NaN      NES          Nintendo   
2                Racing           E      Wii          Nintendo   
3               Shooter         NaN       PC  PUBG Corporation   
4                Sports           E      Wii          Nintendo   
...                 ...         ...      ...               ...   
55787      Role-Playing         NaN       PC         505 Games   
55788         Adventure          RP       PC   Mooneye Studios   
55789         Adventure          RP      PS4   Mooneye Studios   
55790         Adventure          RP     XOne   Mooneye Studios   
55791  Action-Adventure         NaN      PS4           Unknown   

              Developer  VGChartz_Score  Critic_Score  ...  NA_Sales  \
0          Nintendo EAD             NaN           7.7  ...       NaN   
1          Nintendo EAD             NaN          10.0  ...       NaN   
2          Nintendo EAD             NaN           8.2  ...       NaN   
3      PUBG Corporation             NaN           NaN  ...       NaN   
4          Nintendo EAD             NaN           8.0  ...       NaN   
...                 ...             ...           ...  ...       ...   
55787    Lab Zero Games             NaN           NaN  ...       NaN   
55788   Mooneye Studios             NaN           NaN  ...       NaN   
55789   Mooneye Studios             NaN           NaN  ...       NaN   
55790   Mooneye Studios             NaN           NaN  ...       NaN   
55791   Outerloop Games             NaN           NaN  ...       NaN   

       PAL_Sales  JP_Sales  Other_Sales    Year  Last_Update  \
0            NaN       NaN          NaN  2006.0          NaN   
1            NaN       NaN          NaN  1985.0          NaN   
2            NaN       NaN          NaN  2008.0  11th Apr 18   
3            NaN       NaN          NaN  2017.0  13th Nov 18   
4            NaN       NaN          NaN  2009.0          NaN   
...          ...       ...          ...     ...          ...   
55787        NaN       NaN          NaN  2019.0  11th Apr 19   
55788        NaN       NaN          NaN  2019.0  11th Apr 19   
55789        NaN       NaN          NaN  2019.0  11th Apr 19   
55790        NaN       NaN          NaN  2019.0  11th Apr 19   
55791        NaN       NaN          NaN     NaN  12th Apr 19   

                                                     url  status  \
0      http://www.vgchartz.com/game/2667/wii-sports/?...       1   
1      http://www.vgchartz.com/game/6455/super-mario-...       1   
2      http://www.vgchartz.com/game/6968/mario-kart-w...       1   
3      http://www.vgchartz.com/game/215988/playerunkn...       1   
4      http://www.vgchartz.com/game/24656/wii-sports-...       1   
...                                                  ...     ...   
55787  http://www.vgchartz.com/game/224682/indivisibl...       1   
55788  http://www.vgchartz.com/games/game.php?id=2246...       1   
55789  http://www.vgchartz.com/game/224690/l

H0: средняя удовлетворенность критиков спортивными видеоиграм >= 5.0
H1: средняя удовлетворенность критиков спортивными видеоиграм < 5.0

In [3]:
from scipy import stats as st

In [7]:
import numpy as np

In [9]:
#выделяем только те игры, у которых жанр - спорт + где есть оценка критиков
sport = df.loc[df['Genre'] == 'Sports']
sports = sport.loc[sport['Critic_Score'] >= 0]
sports

Rank                                               Name  \
0          1                                         Wii Sports   
4          5                                  Wii Sports Resort   
16        17                                            Wii Fit   
17        18                                       Wii Fit Plus   
68        69                                            FIFA 18   
...      ...                                                ...   
50561  50562  Vancouver 2010 - The Official Video Game of th...   
50730  50731                                      Virtua Tennis   
50731  50732                                      Virtua Tennis   
50840  50841                                          VT Tennis   
51761  51762        World Soccer Winning Eleven 7 International   

                                                basename   Genre ESRB_Rating  \
0                                             wii-sports  Sports           E   
4                                      wii-sports-resort  Sports           E   
16                                               wii-fit  Sports           E   
17                                          wii-fit-plus  Sports           E   
68                                               fifa-18  Sports           E   
...                                                  ...     ...         ...   
50561  vancouver-2010-the-official-video-game-of-the-...  Sports           E   
50730                                      virtua-tennis  Sports           E   
50731                                      virtua-tennis  Sports           E   
50840                                          vt-tennis  Sports           E   
51761        world-soccer-winning-eleven-7-international  Sports           E   

      Platform     Publisher                            Developer  \
0          Wii      Nintendo                         Nintendo EAD   
4          Wii      Nintendo                         Nintendo EAD   
16         Wii      Nintendo                         Nintendo EAD   
17         Wii      Nintendo                         Nintendo EAD   
68         PS4     EA Sports                         EA Vancouver   
...        ...           ...                                  ...   
50561       PC          Sega       Eurocom Entertainment Software   
50730       DC          Sega                             Hitmaker   
50731    NGage          Sega                             Hitmaker   
50840     DSiW  Virtual Toys                         Virtual Toys   
51761       PC        Konami  Konami Computer Entertainment Tokyo   

       VGChartz_Score  Critic_Score  ...  NA_Sales  PAL_Sales  JP_Sales  \
0                 NaN           7.7  ...       NaN        NaN       NaN   
4                 NaN           8.0  ...       NaN        NaN       NaN   
16                NaN           7.9  ...       NaN        NaN       NaN   
17                NaN           8.0  ...       NaN        NaN       NaN   
68                NaN           8.3  ...      1.27       8.64      0.15   
...               ...           ...  ...       ...        ...       ...   
50561             NaN           5.0  ...       NaN        NaN       NaN   
50730             NaN           9.2  ...       NaN        NaN       NaN   
50731             NaN           3.6  ...       NaN        NaN       NaN   
50840             NaN           4.0  ...       NaN        NaN       NaN   
51761             NaN           9.1  ...       NaN        NaN       NaN   

       Other_Sales    Year  Last_Update  \
0              NaN  2006.0          NaN   
4              NaN  2009.0          NaN   
16             NaN  2008.0          NaN   
17             NaN  2009.0          NaN   
68            1.73  2017.0  02nd Apr 18   
...            ...     ...          ...   
50561          NaN  2010.0          NaN   
50730          NaN  2000.0          NaN   
50731          NaN  2003.0          NaN   
50840          NaN  2010.0          NaN   
51761          NaN  2004.0          NaN

In [24]:
alpha = 0.05
result = st.ttest_1samp(sports['Critic_Score'], 5.0)
print(result)

if (result.statistic < 0) & (result.pvalue / 2 < alpha):
    print('Отвергаем нулевую гипотезу, критикам не нравятся спортивные видеоигры')
else:
    print('Не отвергаем нулевую нулевую гипотезу, критикам нравятся спортивные видеоигры')

Ttest_1sampResult(statistic=40.864428307278445, pvalue=4.500149759724238e-186)
Не отвергаем нулевую нулевую гипотезу, критикам нравятся спортивные видеоигры


In [11]:
sport.Critic_Score.mean()

7.291424418604655

Средняя оценка спортивных видеоигр = 7.3, также t-тест показал, что p-value > alpha. Следовательно, критикам нравятся видеоигры

Критикам нравятся больше игры на PC или на PS4?

H0: средняя удовлетворенность критиков PC и PS4 одинаковая
H1: средняя удовлетворенность критиков PC и PS4 различается

In [17]:
PC = df.loc[df['Platform'] == 'PC'].loc[df['Critic_Score'] >= 0]
PS4 = df.loc[df['Platform'] == 'PS4'].loc[df['Critic_Score'] >= 0]

In [18]:
result = st.ttest_ind(PC['Critic_Score'], PS4['Critic_Score'], equal_var=False)
print(result)
if (result.pvalue < alpha):
    print('Отвергаем нулевую гипотезу, критики по-разному относятся к платформам PC и PS4')
else:
    print('Не отвергаем нулевую гипотезу, критики одинаково относятся к платформам PC и PS4')

Ttest_indResult(statistic=-2.7394476056351627, pvalue=0.006931808250254211)
Отвергаем нулевую гипотезу, критики по-разному относятся к платформам PC и PS4


t-тест показал, что критики по-разному относятся к видеогиграм на платформах PC и PS4, более того, поскольку t-критерий отрицательный, то можно сделать вывод, что критики выше оценивают видеоигры на платформе PS4

Критикам больше нравятся стрелялки или стратегии?

In [21]:
df.groupby('Genre')['Critic_Score'].mean()

Genre
Action              6.840234
Action-Adventure    7.693939
Adventure           7.329188
Board Game          5.300000
Education           7.500000
Fighting            7.388415
MMO                 7.810000
Misc                6.885799
Music               7.912000
Party               6.053333
Platform            7.308333
Puzzle              7.200000
Racing              7.091145
Role-Playing        7.461610
Sandbox             9.200000
Shooter             7.286833
Simulation          6.926201
Sports              7.291424
Strategy            7.429268
Visual Novel        8.500000
Name: Critic_Score, dtype: float64

In [ ]:
H0: средняя удовлетворенность критиков стрелялками и стратегиями одинаковая
H1: средняя удовлетворенность критиков стрелялками и стратегиями различается

In [22]:
Shooter = df.loc[df['Genre'] == 'Shooter'].loc[df['Critic_Score'] >= 0]
Strategy = df.loc[df['Genre'] == 'Strategy'].loc[df['Critic_Score'] >= 0]

In [25]:
result = st.ttest_ind(Shooter['Critic_Score'], Strategy['Critic_Score'], equal_var=False)
print(result)
if (result.pvalue < alpha):
    print('Отвергаем нулевую гипотезу, критики по-разному относятся к стрелялкам и стратегиям')
else:
    print('Не отвергаем нулевую гипотезу, критики одинаково относятся к стрелялкам и стратегиям')

Ttest_indResult(statistic=-1.6073949711166526, pvalue=0.10838786414223071)
Не отвергаем нулевую гипотезу, критики одинаково относятся к стрелялкам и стратегиям


t-тест показал, что критики одинаково (вернее, что нет статистической значимости) относятся к стрелкам и стратегиям, более того, но поскольку t-критерий отрицательный, то можно сделать вывод, что критики немного выше оценивают стратегии

In [26]:
Shooter.Critic_Score.mean() < Strategy.Critic_Score.mean()

True

In [27]:
Shooter.Critic_Score.mean() - Strategy.Critic_Score.mean()

-0.14243555246939898

Задание 2
Реализуйте базовую модель логистической регрессии для классификации текстовых сообщенийпо признаку спама. Для этого:

1) Привидите весь текст к нижнему регистру;
2) Удалите мусорные символы;
3) Удалите стоп-слова;
4) Привидите все слова к нормальной форме;
5) Преобразуйте все сообщения в вектора TF-IDF. Вам поможет следующий код:

from sklearn.feature_extraction.text import TfidfVectorizer

tfidf = TfidfVectorizer()
tfidf_matrix = tfidf.fit_transform(df.Message)
names = tfidf.get_feature_names()
tfidf_matrix = pd.DataFrame(tfidf_matrix.toarray(), columns=cname)
Можете поэкспериментировать с параметрами TfidfVectorizer;
6) Разделите данные на тестовые и тренировочные в соотношении 30/70, укажите random_state=42. Используйте train_test_split;
7) Постройте модель логистической регрессии, укажите random_state=42, оцените ее точность на тестовых данных;
8) Опишите результаты при помощи confusion_matrix;
9) Постройте датафрейм, который будет содержать все исходные тексты сообщений, классифицированные неправильно (с указанием фактического и предсказанного).

In [2]:
import re

In [3]:
df = pd.read_csv('spam.csv')
df

Category                                            Message
0         ham  Go until jurong point, crazy.. Available only ...
1         ham                      Ok lar... Joking wif u oni...
2        spam  Free entry in 2 a wkly comp to win FA Cup fina...
3         ham  U dun say so early hor... U c already then say...
4         ham  Nah I don't think he goes to usf, he lives aro...
...       ...                                                ...
5567     spam  This is the 2nd time we have tried 2 contact u...
5568      ham               Will ü b going to esplanade fr home?
5569      ham  Pity, * was in mood for that. So...any other s...
5570      ham  The guy did some bitching but I acted like i'd...
5571      ham                         Rofl. Its true to its name

[5572 rows x 2 columns]

In [4]:
import nltk

In [5]:
nltk.download('stopwords')

[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\Ксения\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


True

In [6]:
nltk.download('wordnet')

[nltk_data] Downloading package wordnet to
[nltk_data]     C:\Users\Ксения\AppData\Roaming\nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


True

In [7]:
from nltk.corpus import stopwords
stopwords_set = set(stopwords.words('english'))

In [8]:
from nltk.stem import WordNetLemmatizer
wordnet_lemmatizer = WordNetLemmatizer()

In [9]:
df['Message'] = df['Message'].str.lower()  #приводим к нижнему регистру
df['Message'] = df['Message'].replace('[\W_]+',' ', regex = True) # убираем все служебные символы
df['Message'] = df['Message'].apply(lambda x: [word for word in x.split() if word not in stopwords_set]) # убираем стоп-слова
df['Message'] = df['Message'].apply(lambda x: [wordnet_lemmatizer.lemmatize(word) for word in x]) #приводим слова к словарной форме

In [10]:
from sklearn.feature_extraction.text import TfidfVectorizer

In [11]:
df['Message'] = df['Message'].astype(str)

In [12]:
tfidf = TfidfVectorizer()
tfidf_matrix = tfidf.fit_transform(df.Message)

In [13]:
names = tfidf.get_feature_names()
tfidf_matrix = pd.DataFrame(tfidf_matrix.toarray(), columns=names)

In [14]:
data = tfidf_matrix

In [15]:
Category = df['Category']

In [16]:
data['Category'] = Category

In [17]:
from sklearn.model_selection import train_test_split
data_train, data_test = train_test_split(data, test_size=0.3, random_state=42)

In [18]:
y_train, y_test = data_train['Category'], data_test['Category']
del data_train['Category']
del data_test['Category']

In [19]:
from sklearn.linear_model import LogisticRegression

In [20]:
clf = LogisticRegression(random_state=42).fit(data_train, y_train)

In [21]:
y_pred = clf.predict(data_test)

In [22]:
y_pred

array(['ham', 'ham', 'ham', ..., 'ham', 'spam', 'ham'], dtype=object)

In [23]:
from sklearn.metrics import accuracy_score

In [24]:
accuracy_score(y_test, y_pred)

0.958732057416268

Ошибка составляет около 5%

In [25]:
from sklearn.metrics import confusion_matrix

In [26]:
confusion_matrix(y_test, y_pred)

array([[1445,    3],
       [  66,  158]], dtype=int64)

1445, которые были true, определены как true.
158, которые были false, определены как false.
т.е. верны определены (1445 +158)/1672 = 0.958732057


66, которые были false, определены как true
3, которые были true, определены как false
Неверны определены (66+3)/1672 = 0.041267943

In [42]:
data_test

00       000  000pes  008704050406  0089  0121  01223585236  \
3245  0.0  0.000000     0.0           0.0   0.0   0.0          0.0   
944   0.0  0.000000     0.0           0.0   0.0   0.0          0.0   
1044  0.0  0.000000     0.0           0.0   0.0   0.0          0.0   
2484  0.0  0.000000     0.0           0.0   0.0   0.0          0.0   
812   0.0  0.000000     0.0           0.0   0.0   0.0          0.0   
...   ...       ...     ...           ...   ...   ...          ...   
2505  0.0  0.000000     0.0           0.0   0.0   0.0          0.0   
2525  0.0  0.000000     0.0           0.0   0.0   0.0          0.0   
4975  0.0  0.000000     0.0           0.0   0.0   0.0          0.0   
650   0.0  0.566635     0.0           0.0   0.0   0.0          0.0   
4463  0.0  0.000000     0.0           0.0   0.0   0.0          0.0   

      01223585334  0125698789   02  ...  zhong  zindgi  zoe  zogtorius  zoom  \
3245          0.0         0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   
944           0.0         0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   
1044          0.0         0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   
2484          0.0         0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   
812           0.0         0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   
...           ...         ...  ...  ...    ...     ...  ...        ...   ...   
2505          0.0         0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   
2525          0.0         0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   
4975          0.0         0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   
650           0.0         0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   
4463          0.0         0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   

      zouk  zyada   èn   ú1  〨ud  
3245   0.0    0.0  0.0  0.0  0.0  
944    0.0    0.0  0.0  0.0  0.0  
1044   0.0    0.0  0.0  0.0  0.0  
2484   0.0    0.0  0.0  0.0  0.0  
812    0.0    0.0  0.0  0.0  0.0  
...    ...    ...  ...  ...  ...  
2505   0.0    0.0  0.0  0.0  0.0  
2525   0.0    0.0  0.0  0.0  0.0  
4975   0.0    0.0  0.0  0.0  0.0  
650    0.0    0.0  0.0  0.0  0.0  
4463   0.0    0.0  0.0  0.0  0.0  

[1672 rows x 8048 columns]

In [27]:
data_new = data_test
data_new['y_fact'] = y_test
data_new['y_pred'] = y_pred

data_new

C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until


00       000  000pes  008704050406  0089  0121  01223585236  \
3245  0.0  0.000000     0.0           0.0   0.0   0.0          0.0   
944   0.0  0.000000     0.0           0.0   0.0   0.0          0.0   
1044  0.0  0.000000     0.0           0.0   0.0   0.0          0.0   
2484  0.0  0.000000     0.0           0.0   0.0   0.0          0.0   
812   0.0  0.000000     0.0           0.0   0.0   0.0          0.0   
...   ...       ...     ...           ...   ...   ...          ...   
2505  0.0  0.000000     0.0           0.0   0.0   0.0          0.0   
2525  0.0  0.000000     0.0           0.0   0.0   0.0          0.0   
4975  0.0  0.000000     0.0           0.0   0.0   0.0          0.0   
650   0.0  0.566635     0.0           0.0   0.0   0.0          0.0   
4463  0.0  0.000000     0.0           0.0   0.0   0.0          0.0   

      01223585334  0125698789   02  ...  zoe  zogtorius  zoom  zouk  zyada  \
3245          0.0         0.0  0.0  ...  0.0        0.0   0.0   0.0    0.0   
944           0.0         0.0  0.0  ...  0.0        0.0   0.0   0.0    0.0   
1044          0.0         0.0  0.0  ...  0.0        0.0   0.0   0.0    0.0   
2484          0.0         0.0  0.0  ...  0.0        0.0   0.0   0.0    0.0   
812           0.0         0.0  0.0  ...  0.0        0.0   0.0   0.0    0.0   
...           ...         ...  ...  ...  ...        ...   ...   ...    ...   
2505          0.0         0.0  0.0  ...  0.0        0.0   0.0   0.0    0.0   
2525          0.0         0.0  0.0  ...  0.0        0.0   0.0   0.0    0.0   
4975          0.0         0.0  0.0  ...  0.0        0.0   0.0   0.0    0.0   
650           0.0         0.0  0.0  ...  0.0        0.0   0.0   0.0    0.0   
4463          0.0         0.0  0.0  ...  0.0        0.0   0.0   0.0    0.0   

       èn   ú1  〨ud  y_fact  y_pred  
3245  0.0  0.0  0.0     ham     ham  
944   0.0  0.0  0.0     ham     ham  
1044  0.0  0.0  0.0     ham     ham  
2484  0.0  0.0  0.0     ham     ham  
812   0.0  0.0  0.0     ham     ham  
...   ...  ...  ...     ...     ...  
2505  0.0  0.0  0.0     ham     ham  
2525  0.0  0.0  0.0    spam    spam  
4975  0.0  0.0  0.0     ham     ham  
650   0.0  0.0  0.0    spam    spam  
4463  0.0  0.0  0.0     ham     ham  

[1672 rows x 8050 columns]

In [34]:
for strings in data_new:
    if data_new['y_fact']!= data_new['y_pred']:
        print(strings)

ValueError: The truth value of a Series is ambiguous. Use a.empty, a.bool(), a.item(), a.any() or a.all().